adr/0002-masking-configuration-API.md

```diff
@@ -0,0 +1,199 @@
+# 2. API design for JSON masking configurations
+
+Date: 23-01-2024
+
+## Status
+
+Accepted by:
+
+- [Breus](https://github.com/Breus)
+- [gavlyukovskiy](https://github.com/gavlyukovskiy)
+- [donavdey](https://github.com/donavdey)
+
+## Context
+
+Before releasing version 1.x, we want to make the JSON masking configuration API as extensible as possible while
+providing predictable defaults. The current API was influenced by performance considerations which no longer apply due
+to certain optimizations done to the implementation. This resulted in a suboptimal API both in terms of extensibility
+and in terms of default behaviour.
+
+## Decisions
+
+### Change the default masking behaviors for all JSON types
+
+By default we
+1. We will not make a masking decision based on type. 
+    The main reason is that masking is explicitly requested by the user by specifying the target keys (either block
+    or allow) and it would be unexpected that we do not mask an explicitly requested key because it's a number and 
+    number masking is not enabled by default.
+2. We will enable length obfuscation by default as length on its own might leak information about the original value.
+3. We will mask all types into strings.
+    This is done to keep masking consistent across types and to make it apparent that the value was masked (while `"***"`
+    is a good indicator for strings, the `888` or `0` for numbers, or `false` for booleans is not)
+4. We will use different masks for each original type in to keep information about the original type, the strings will 
```

Comment:
Nit: remove `in` in ` type in to keep`

Reply:
```suggestion
4. We will use different masks for each original type to keep information about the original type, the strings will 
```
FYI GitHub has suggestions feature (icon of file with +-)
<img width="309" alt="image" src="https://github.com/Breus/json-masker/assets/15277543/939db7d8-1b64-43a3-bf32-f48deac44a73">

Reply:
Ah yes good suggestion. In Gitlab internally that doesn't work so I unlearned myself this 😋